lib/vssh/libssh2.c

```diff
@@ -694,25 +694,27 @@ static CURLcode ssh_force_knownhost_key_type(struct connectdata *conn)
     while(!libssh2_knownhost_get(sshc->kh, &store, store)) {
       /* For non-standard ports, the name will be enclosed in */
       /* square brackets, followed by a colon and the port */
-      if(store->name[0] == '[') {
-        kh_name_end = strstr(store->name, "]:");
-        if(!kh_name_end) {
-          infof(data, "Invalid host pattern %s in %s\n",
-                store->name, data->set.str[STRING_SSH_KNOWNHOSTS]);
-          continue;
-        }
-        port = atoi(kh_name_end + 2);
-        if(kh_name_end && (port == conn->remote_port)) {
-          kh_name_size = strlen(store->name) - 1 - strlen(kh_name_end);
-          if(strncmp(store->name + 1, conn->host.name, kh_name_size) == 0) {
-            found = true;
-            break;
+      if(store->name){
+        if(store->name[0] == '[') {
+          kh_name_end = strstr(store->name, "]:");
+          if(!kh_name_end) {
+            infof(data, "Invalid host pattern %s in %s\n",
+                  store->name, data->set.str[STRING_SSH_KNOWNHOSTS]);
+            continue;
+          }
+          port = atoi(kh_name_end + 2);
+          if(kh_name_end && (port == conn->remote_port)) {
+            kh_name_size = strlen(store->name) - 1 - strlen(kh_name_end);
+            if(strncmp(store->name + 1, conn->host.name, kh_name_size) == 0) {
+              found = true;
+              break;
+            }
           }
         }
-      }
-      else if(strcmp(store->name, conn->host.name) == 0) {
-        found = true;
-        break;
+        else if(strcmp(store->name, conn->host.name) == 0) {
+          found = true;
+          break;
+        }
```

Comment:
If libssh2_knownhost_get() returns zero, it means it has found a match. Shouldn't that mean something even if we don't get `name` set to anything?

Reply:
had the same thought but unless I am mistaken store->name would never match conn->host.name ... unsure what condition to check.

Reply:
@SantinoKeupp

Reply:
If it can't match a host name, then curl can't "force" a specify hostkey_method. That doesn't mean that the host is unknown. Modern known host files do not contain the host name in a plain version so most entries will not be possible to match this way. I don't think the connection should fail because if it, it just host fail to set a preferred method.

Reply:
I will note that this problem I mention was brought by #4747 already, I very clearly didn't review it correctly at the time.

Reply:
I understand - my default (wrong) expectation was we needed to explicitly match host name  ... will adjust.